cart/src/endpoints/handlers.js

```diff
@@ -0,0 +1,281 @@
+/**
+ * These are the handlers for the endpoints for the cart.
+ * It's the code that actually contains the logic for each endpoint.
+ */
+
+'use strict';
+
+const logger = require('../utils/logger');
+const Model = require('../models').Model;
+
+/**
+ * The handler functions for all endpoints defined for the cart
+ */
+class Handlers {
+  constructor() {
+    this.model = new Model();
+    this.logger = logger;
+  }
+
+  /**
+   * Check if properties are present in an object, returns an object with a boolean value.
+   * When a property is missing, it also returns the missing property.
+   * @static
+   * @param {array} proplist - an array of properties that you want to check
+   * @param {object} obj - the object to check
+   */
+  static propsPresent(proplist, obj) {
+    for (let prop of proplist) {
+      if (!(prop in obj)) {
+        return {valid: false, missing: prop};
+      }
+    }
+    return {valid: true};
+  }
+
+  /**
+   * Log a request using Winston
+   * @param {Hapi.request} req - the request object
+   */
+  logRequest(req) {
+    this.logger.info(`HTTP ${req.method} ${req.path}`)
+  }
+
+  /**
+   * Add a product to a cart.
+   * @async
+   * @param {Hapi.request} req - the request object
+   * @param {object} rep - the response toolkit (Hapi.h)
+   */
+  async addProduct(req, rep) {
+    this.logRequest(req);
+
+    // Check if request contains a body
+    if (!req.payload) {
+      return rep.response({message: "Body cannot be empty."}).code(400);
+    }
+
+    // Check if request body contains the required values
+    const isValid = Handlers.propsPresent(['pid', 'amount_in_cart'], req.payload);
+    if (!isValid.valid) {
+      return rep.response({message: `${isValid.missing} not specified.`}).code(400);
+    }
+
+    this.logger.debug(`\tHandler: Adding product ${JSON.stringify(req.payload)} to cart ${req.params.id}`);
+
+    try {
+      // Add the product to the cart
+      const res = await this.model.addProduct(req.params.id, req.payload);
+      this.logger.debug(`\tResult: ${JSON.stringify(res)}`);
+
+      // Return what was added
+      return rep.response({message: "Product added to cart.", data: res}).code(201);
+    }
+
+    // Catch any database errors (e.g. product not found) and return the appropriate response
+    catch (err) {
+      if (err.constraint === "products_in_cart_pid_fkey") {
+        let message = 'Product does not exist.';
+        this.logger.debug(`\t${message}`);
+        return rep.response({message: message}).code(400);
+      }
+      else if (err.constraint === "carts_uid_fkey") {
+        let message = 'User does not exist.';
+        this.logger.debug(`\t${message} -- ${err.detail}`);
+        return rep.response({message: message}).code(400);
+      }
+      else if (err.constraint === "products_in_cart_pkey") {
+        let message = 'Product already present in cart.';
+        this.logger.debug(`\t${message} -- ${err.detail}`);
```

Comment:
`err.detail` is a detailed message regarding the error, it's created by knex iirc. It's not something I create, but it makes for a more useful error message. It also already includes the id that was requested. `err.detail` gives you e.g. "Key (cartid, pid)=(1, 1) already exists."

What would you have in mind for a stack trace?

Reply:
I was thinking about printing the entire stacktrace, but let's keep it as is for now since this error would be query error and not the code error and the stacktrace won't be helpful here. Also since err.detail adds necessary info, i think it should be good for now